cmd/util/printer/job_progress_printer.go

```diff
@@ -0,0 +1,374 @@
+package printer
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"strings"
+	"time"
+
+	"github.com/bacalhau-project/bacalhau/cmd/util"
+	"github.com/bacalhau-project/bacalhau/cmd/util/flags/cliflags"
+	"github.com/bacalhau-project/bacalhau/cmd/util/output"
+	"github.com/bacalhau-project/bacalhau/pkg/bacerrors"
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+	"github.com/bacalhau-project/bacalhau/pkg/publicapi/apimodels"
+	clientv2 "github.com/bacalhau-project/bacalhau/pkg/publicapi/client/v2"
+	"github.com/bacalhau-project/bacalhau/pkg/util/idgen"
+	"github.com/jedib0t/go-pretty/v6/table"
+	"github.com/jedib0t/go-pretty/v6/text"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/samber/lo"
+	"github.com/spf13/cobra"
+)
+
+const PrintoutCanceledButRunningNormally string = "printout canceled but running normally"
+
+type JobProgressPrinter struct {
+	client          clientv2.API
+	runtimeSettings *cliflags.RunTimeSettings
+}
+
+type jobProgressEvent struct {
+	jobID       string
+	occurred    time.Time
+	executionID string
+	eventTopic  models.EventTopic
+	event       models.Event
```

Comment:
aren't topic and occured already part of the event?

Reply:
So occurred, inside job history has a logic, where if the event timestamp is not present it returns the history timestamp. Hence tracking it separately.